lib/live-preview/handleStoryblokMessage.ts

```diff
@@ -0,0 +1,75 @@
+import type { ISbStoryData } from "@storyblok/js";
+
+let timeout: NodeJS.Timeout;
+/**
+ * Our current tests indicate that the post request section
+ * is the bottleneck in terms of time consumption.
+ * We should explore alternative methods to optimize
+ * this process and improve efficiency.
+ */
+export async function handleStoryblokMessage(event: {
+  action: string;
+  story: ISbStoryData;
+}) {
+  const { action, story } = event || {};
+
+  if (action === "input" && story) {
+    // Debounce the getNewHTMLBody function
+    const debouncedGetNewHTMLBody = async () => {
+      const newBody = await getNewHTMLBody(story);
+      const currentBody = document.body as HTMLElement;
+      if (newBody.outerHTML === currentBody.outerHTML) return;
+      // Get current focused element in Storyblok
+      const focusedElem = document.querySelector('[data-blok-focused="true"]');
+      updateDOMWithNewBody(currentBody, newBody, focusedElem);
+    };
+    const debounceDelay = 1000; // Adjust the delay as needed
+    clearTimeout(timeout);
+    timeout = setTimeout(debouncedGetNewHTMLBody, debounceDelay);
+  }
+
+  if (["published", "change"].includes(event?.action)) {
+    location.reload();
+  }
+}
+
+function updateDOMWithNewBody(
+  currentBody: HTMLElement,
+  newBody: HTMLElement,
+  focusedElem: Element | null
+) {
+  if (focusedElem) {
+    //Get the [data-blok-uid] of the focused element in storyblok
+    const focusedElementID = focusedElem.getAttribute("data-blok-uid");
+    //Now find the same element by above [data-blok-uid] in our new virtual HTML page
+    const newDomFocusElem = newBody.querySelector(
+      `[data-blok-uid="${focusedElementID}"]`
+    );
+    if (newDomFocusElem) {
+      // Add the [data-blok-focused] attribute to the above element
+      newDomFocusElem.setAttribute("data-blok-focused", "true");
+      // console.log("Doing partial replace");
+      focusedElem.replaceWith(newDomFocusElem);
+    }
+  } else {
+    // console.log("Doing full replace");
+    currentBody.replaceWith(newBody);
+  }
+}
+
+async function getNewHTMLBody(story: ISbStoryData) {
+  const result = await fetch(location.href, {
```

Comment:
_(Maybe for the next iteration / Worth fiddling with in the meantime)_
This function can currently handle only the happy path.
I'd try to think about what can go wrong here, so that we can make it more robust and handle errors.
For instance, the `fetch` might fail for a variety of reasons (50x, 405, etc.).
How do we want to handle such cases?